File: sample-app-java/app/helper/FakeServiceClient.java
package helper;

import akka.actor.ActorSystem;
import com.fasterxml.jackson.databind.JsonNode;
import data.FutureUtil;
import data.Response;
import play.api.libs.json.JsValue;
import play.libs.F;
import play.libs.Json;

import javax.inject.Inject;

public class FakeServiceClient {
  private final data.FakeServiceClient delegate;

  @Inject
  public FakeServiceClient(ActorSystem actorSystem) {
    delegate = new data.FakeServiceClient(new FutureUtil(actorSystem));
  }

  public F.Promise<Response> fakeRemoteCallFast(String id) {
    return F.Promise.wrap(delegate.fakeRemoteCallFast(id));
  }

  public F.Promise<Response> fakeRemoteCallMedium(String id) {
    return F.Promise.wrap(delegate.fakeRemoteCallMedium(id));
  }

  public F.Promise<Response> fakeRemoteCallSlow(String id) {
    return F.Promise.wrap(delegate.fakeRemoteCallSlow(id));
  }

  public F.Promise<Response> fakeRemoteCall(String id, long delayInMillis) {
    return F.Promise.wrap(delegate.fakeRemoteCall(id, delayInMillis));
  }

  public F.Promise<JsonNode> fakeRemoteCallJsonFast(String id) {
    return toJsonNode(F.Promise.wrap(delegate.fakeRemoteCallJsonFast(id)));
  }

  public F.Promise<JsonNode> fakeRemoteCallJsonMedium(String id) {
    return toJsonNode(F.Promise.wrap(delegate.fakeRemoteCallJsonMedium(id)));
  }

  public F.Promise<JsonNode> fakeRemoteCallJsonSlow(String id) {
    return toJsonNode(F.Promise.wrap(delegate.fakeRemoteCallJsonSlow(id)));
  }

  public F.Promise<JsonNode> fakeRemoteCallJson(String id, long delayInMillis) {
    return toJsonNode(F.Promise.wrap(delegate.fakeRemoteCallJson(id, delayInMillis)));
  }

  private F.Promise<JsonNode> toJsonNode(F.Promise<JsValue> jsValuePromise) {
    return jsValuePromise.map(jsValue -> Json.parse(jsValue.toString()));
  }

  public F.Promise<Response> fakeRemoteCallErrorFast(String id) {
    return F.Promise.wrap(delegate.fakeRemoteCallErrorFast(id));
  }

  public F.Promise<Response> fakeRemoteCallErrorMedium(String id) {
    return F.Promise.wrap(delegate.fakeRemoteCallErrorMedium(id));
  }

  public F.Promise<Response> fakeRemoteCallErrorSlow(String id) {
    return F.Promise.wrap(delegate.fakeRemoteCallErrorSlow(id));
  }

  public F.Promise<Response> fakeRemoteCallError(String id, long delayInMillis) {
    return F.Promise.wrap(delegate.fakeRemoteCallError(id, delayInMillis));
  }
}

File: sample-app-common/src/main/scala/data/FakeServiceClient.scala
package data

import play.api.libs.concurrent.Execution.Implicits._
import play.api.libs.json.{Json, JsValue}

import scala.concurrent.Future
import scala.util.Random
import data.Response._

/**
 * A client that represents fake calls to remote backend services.
 */
class FakeServiceClient(futureUtil: FutureUtil) {

  import data.FakeServiceClient._

  def fakeRemoteCallFast(id: String): Future[Response] = fakeRemoteCall(id, FAST_RESPONSE_TIME_IN_MILLIS)
  def fakeRemoteCallMedium(id: String): Future[Response] = fakeRemoteCall(id, MEDIUM_RESPONSE_TIME_IN_MILLIS)
  def fakeRemoteCallSlow(id: String): Future[Response] = fakeRemoteCall(id, SLOW_RESPONSE_TIME_IN_MILLIS)

  def fakeRemoteCallJsonFast(id: String): Future[JsValue] = fakeRemoteCallJson(id, FAST_RESPONSE_TIME_IN_MILLIS)
  def fakeRemoteCallJsonMedium(id: String): Future[JsValue] = fakeRemoteCallJson(id, MEDIUM_RESPONSE_TIME_IN_MILLIS)
  def fakeRemoteCallJsonSlow(id: String): Future[JsValue] = fakeRemoteCallJson(id, SLOW_RESPONSE_TIME_IN_MILLIS)
  def fakeRemoteCallJson(id: String, delayInMillis: Long): Future[JsValue] = fakeRemoteCall(id, delayInMillis).map(Json.toJson(_))

  def fakeRemoteCallErrorFast(id: String): Future[Response] = fakeRemoteCallError(id, FAST_RESPONSE_TIME_IN_MILLIS)
  def fakeRemoteCallErrorMedium(id: String): Future[Response] = fakeRemoteCallError(id, MEDIUM_RESPONSE_TIME_IN_MILLIS)
  def fakeRemoteCallErrorSlow(id: String): Future[Response] = fakeRemoteCallError(id, SLOW_RESPONSE_TIME_IN_MILLIS)

  def fakeRemoteCallError(id: String, delayInMillis: Long): Future[Response] = {
    fakeRemoteCall(id, delayInMillis).map(response => throw FakeRemoteCallException(response))
  }

  def fakeRemoteCall(id: String, delayInMillis: Long): Future[Response] = {
    val randomJitter = new Random().nextInt(delayInMillis.toInt).toLong
    val delay = delayInMillis + randomJitter

    val fakeJsonResponse = Response(id, delay)
    futureUtil.timeout(fakeJsonResponse, delay)
  }
}

object FakeServiceClient {
  val FAST_RESPONSE_TIME_IN_MILLIS = 5
  val MEDIUM_RESPONSE_TIME_IN_MILLIS = 500
  val SLOW_RESPONSE_TIME_IN_MILLIS = 3000
}

case class FakeRemoteCallException(response: Response) extends RuntimeException(s"""Error in "${response.id}" after ${response.delay} ms""")


